src/daft-local-execution/src/dynamic_batching/latency_constrained.rs

```diff
@@ -0,0 +1,292 @@
+        morsel_size_req: Option<&MorselSizeRequirement>,
+    ) -> Self {
+        let (min, max) = match morsel_size_req {
+            Some(MorselSizeRequirement::Strict(v)) => (0, *v),
+            Some(MorselSizeRequirement::Flexible(min, max)) => (*min, *max),
+            None => (0, 128 * 1024),
+        };
+        Self {
+            target_batch_latency: target_decoding_latency,
+            latency_tolerance: latency_slack,
+            step_size_alpha,
+            correction_delta,
+            min_batch_size: min,
+            max_batch_size: max,
+        }
+    }
+    pub fn with_target_batch_latency(mut self, target_batch_latency: Duration) -> Self {
+        self.target_batch_latency = target_batch_latency;
+        self
+    }
+    pub fn with_latency_tolerance(mut self, latency_tolerance: Duration) -> Self {
+        self.latency_tolerance = latency_tolerance;
+        self
+    }
+    pub fn with_correction_delta(mut self, correction_delta: usize) -> Self {
+        self.correction_delta = correction_delta;
+        self
+    }
+
+    pub fn with_step_size(mut self, step_size: usize) -> Self {
+        self.step_size_alpha = step_size;
+        self
+    }
+}
+
+impl Default for LatencyConstrainedBatchingStrategy {
+    fn default() -> Self {
+        Self {
+            target_batch_latency: Duration::from_millis(5000),
+            latency_tolerance: Duration::from_millis(1000),
+            step_size_alpha: 1024,
+            correction_delta: 10,
+            min_batch_size: 1,
+            max_batch_size: 128 * 1024,
+        }
+    }
+}
+
+pub struct LatencyConstrainedBatchingState {
+    current_batch_size: usize,
+    search_low: usize,
+    search_high: usize,
+}
+
+impl LatencyConstrainedBatchingState {
+    pub fn new(initial_batch_size: usize, min: usize, max: usize) -> Self {
+        Self {
+            current_batch_size: initial_batch_size,
+            search_low: min,
+            search_high: max,
+        }
+    }
+}
+impl BatchingState for LatencyConstrainedBatchingState {}
+
+impl BatchingStrategy for LatencyConstrainedBatchingStrategy {
+    type State = LatencyConstrainedBatchingState;
+
+    fn make_state(&self) -> Self::State {
+        log::debug!(
+            "[{}] Initializing state with search space [1, 256]",
+            std::thread::current().name().unwrap_or("unknown")
+        );
+        // start off with a small search space (1 - 256)
+        LatencyConstrainedBatchingState::new(self.min_batch_size, self.min_batch_size, 256)
+    }
+
+    fn initial_batch_size(&self) -> usize {
+        self.min_batch_size
+    }
+
+    fn calculate_new_batch_size(
+        &self,
+        state: &mut Self::State,
+        batch: Vec<(std::sync::Arc<dyn RuntimeStats>, usize, Duration)>,
+    ) -> usize {
+        let latency = avg_latency(batch.iter().map(|(_, _, latency)| *latency));
+        let batch_size = avg_batch_size(batch.iter().map(|(_, batch_size, _)| *batch_size));
```

Comment:
note: this had to be slightly modified from the original paper because we can't guarantee that we're recalculating the batch size after every single morsel is sent. We instead just do it periodically. As such, we need to do it on the averages instead of the absolute values.